agents: add test-pyramid skill for unit/E2E test rebalancing (#1199)
kubernetes-sigs/agent-sandbox: .agents/skills/test-pyramid/SKILL.md

---
name: test-pyramid
description: Analyze the repo's unit and E2E tests and propose rebalancing toward a test pyramid — which E2E tests (or assertions inside them) can be covered by unit tests, which unit-level gaps genuinely need E2E coverage, and where coverage is duplicated. Use when the user asks about test pyramid, test rebalancing, "should this be an e2e or unit test", E2E-to-unit migration, or slow/flaky E2E suites that might shrink.
---

# Test Pyramid Analysis

Goal: **anything that can be tested as a unit test should be a unit test; only what genuinely needs a real cluster should be E2E.** The end state is a pyramid — many fast unit tests, few E2E tests. This skill produces an evidence-backed rebalancing report; it does NOT move tests itself unless the user asks afterward.

## Phase 1 — Inventory

Enumerate both layers and count actual test functions (not just files), so the report can show the pyramid shape numerically:

> - **Unit tests:** `git ls-files '*_test.go' | grep -v '^test/'` plus Python unit tests under `clients/python/**/tests/`. Count `func Test...` per package (`grep -c '^func Test'`).

Author: Fixed in 81200ae — the inventory now points at the actual `clients/python/**/test/unit/` layout (with a discovery command, since there are two such directories) and counts `def test_` for Python. Applied locally, no commit-suggestion button.

> - **Unit tests:** `git ls-files '*_test.go' | grep -v '^test/'` plus Python unit tests under `clients/python/**/test/unit/` (discover with `git ls-files 'clients/python/*' | grep '/test/unit/'` — there is more than one such directory). Count `func Test...` per package (`grep -c '^func Test'`) and `def test_` for Python.

Reviewer: `grep -v '^test/'` excludes `test/e2e/framework/*`, which are unit tests for framework components. 

```suggestion
- **Unit tests:** `git ls-files '*_test.go' | grep -v '^test/e2e/'` plus `git ls-files 'test/e2e/framework/*_test.go'` and Python unit tests under `clients/python/**/test/unit/` (discover with `git ls-files 'clients/python/*' | grep '/test/unit/'` — there is more than one such directory). Count `func Test...` per package (`grep -c '^func Test'`) and `def test_` for Python.
```

Author: Fixed in 21c8784 — applied your suggestion locally (no commit-suggestion button): unit inventory now uses `grep -v '^test/e2e/'` plus `test/e2e/framework/*_test.go`. Also excluded `test/e2e/framework/` from the E2E enumeration on the next line so those tests are not double-counted at both layers.

> - **E2E / system tests:** `test/e2e/**/*_test.go` excluding `test/e2e/framework/` (kind-cluster E2E, incl. `test/e2e/extensions/`), `test/e2e/clients/python/` (SDK E2E), `dev/tools/test-migration.py` (upgrade/rollback), `test/stress/` (load). Count `func Test...` and, for table-driven E2E, the sub-scenarios.

Reviewer: The phrasing is ambiguous, it makes `test/e2e/clients/python/`, `dev/tools/test-migration.py`, and `test/stress/` look like they are part of the "excluding" clause, when they are actually additional E2E/system test suites to include alongside `test/e2e/*_test.go`. Additionally, using `test/e2e/*_test.go` is cleaner and git-safe, and we should mention counting `def test_` for Python/migration tests.

```suggestion
- **E2E / system tests:** `test/e2e/*_test.go` (excluding `test/e2e/framework/`), plus `test/e2e/clients/python/` (SDK E2E), `dev/tools/test-migration.py` (upgrade/rollback), and `test/stress/` (load). Count `func Test...` (and `def test_` for Python / migration), and for table-driven E2E, the sub-scenarios.
```

Author: Fixed in fee7046 — applied your suggestion locally:
- `test/e2e/*_test.go` (git-safe glob) instead of `test/e2e/**/*_test.go`
- `test/e2e/extensions/`, `test/e2e/clients/python/`, `dev/tools/test-migration.py`, and `test/stress/` are now clearly separate `plus` items rather than reading as part of the `excluding test/e2e/framework/` clause
- Added `def test_` counting for Python/migration tests

Also took the opportunity to broaden the Python unit-test discovery to cover `clients/integrations/deepagents/tests/unit/` and `clients/integrations/mcp-server/tests/unit/`, which were added to main after this PR was first filed.

- **Keep as E2E:** tests dominated by cluster-physics assertions. Do not propose unit-testing scheduling, pod readiness, real router traffic, webhook round-trips, or migration/upgrade behavior — a fake client proves nothing there.
- **Unit → E2E promotion (rare, keep the pyramid in mind):** unit tests that mock so much they only test the mock (assert the seam is meaningfully exercised), or critical user journeys (create claim → adopt → route traffic → shutdown) with no E2E smoke path at all. Prefer ONE thin journey test over per-feature E2E.
- **Redundant coverage:** the same seam asserted at both layers with the same inputs — keep the unit test, list the E2E assertion as prunable.

## Phase 4 — Verify before reporting

Every suggestion must survive these checks (drop or downgrade to "uncertain" if not):

- Name the exact seam (file:line of the function) and confirm it is callable without a cluster — no unexported entanglement with live clients that fake clients can't satisfy.

> - Check a unit test for that seam doesn't already exist (`grep` the seam name across `*_test.go`); if it exists, the finding is "redundant E2E assertion", not "missing unit test".

Reviewer: A broad `grep` across `*_test.go` for the seam name can match comments or unrelated identifiers

```suggestion
- Check a unit test for that seam doesn't already exist (use package-qualified symbol search across `*_test.go` and verify the test invokes the function); if it exists, the finding is "redundant E2E assertion", not "missing unit test".
```

Author: Fixed in 21c8784 — applied your suggestion locally: the verify step now calls for a package-qualified symbol search and confirming the test actually invokes the function, instead of a bare grep of the seam name.

- Confirm the E2E test would still have a reason to exist after extraction, or explicitly state it can be deleted and what residual smoke coverage (if any) replaces it.
- Beware behaviors that LOOK like logic but are cluster-coupled: informer cache timing, conversion-webhook storage effects, owner-reference garbage collection, Prow retest semantics, anything the migration test covers. When unsure, classify as keep-E2E and say why.

## Phase 5 — Report (inline only; write no files)

1. **Pyramid snapshot:** test-function counts per layer (unit / E2E / migration / stress) now vs. after adopting all suggestions, plus CI-time estimate if available.
2. **E2E → unit table:** E2E test (file:line) | assertions to extract | seam (file:line) | proposed unit test location + shape (table-driven case to add vs. new test) | what remains of the E2E test (thinned / deleted).
3. **Unit → E2E table** (expect this to be short — the pyramid demands it): gap | why unit level cannot cover it | proposed E2E home (existing file to extend before new file).
4. **Redundant coverage list.**

> 5. **Top 5 quick wins** ranked by (CI time + flake history removed) vs. effort — cross-reference `dev/tools/flake-report` output or open `kind/flake` issues if available, since migrating a flaky E2E assertion to a unit test is the highest-value move.

Author: Fixed in 81200ae — the quick-wins step now leads with sources that exist on main (open kind/flake issues, TestGrid presubmit history) and mentions `dev/tools/flake-report` only conditionally ("if that tool exists in your checkout"). Applied locally.
